src/ivectorbin/ivector-extract-dense.cc

```diff
@@ -0,0 +1,210 @@
+  
+    SequentialBaseFloatMatrixReader feature_reader(feature_rspecifier);
+    RandomAccessPosteriorReader posterior_reader(posterior_rspecifier);
+    BaseFloatMatrixWriter ivectors_writer(ivectors_wspecifier);
+  
+    for (; !feature_reader.Done(); feature_reader.Next()) {
+      std::string utt = feature_reader.Key();
+      if (!posterior_reader.HasKey(utt)) {
+        KALDI_WARN << "No posteriors for utterance " << utt;
+        num_err++;
+        continue;
+      }
+      const Matrix<BaseFloat> &mat = feature_reader.Value();
+      Posterior posterior = posterior_reader.Value(utt);
+      
+      if (static_cast<int32>(posterior.size()) != mat.NumRows()) {
+        KALDI_WARN << "Size mismatch between posterior " << posterior.size()
+                   << " and features " << mat.NumRows() << " for utterance "
+                   << utt;
+        num_err++;
+        continue;
+      }
+
+      double *auxf_ptr = (compute_objf_change ? &tot_auxf_change : NULL );
+
+      double this_t = opts.acoustic_weight * TotalPosterior(posterior),
+          max_count_scale = 1.0;
+      if (opts.max_count > 0 && this_t > opts.max_count) {
+        max_count_scale = opts.max_count / this_t;
+        KALDI_LOG << "Scaling stats for utterance " << utt << " by scale "
+                  << max_count_scale << " due to --max-count="
+                  << opts.max_count;
+        this_t = opts.max_count;
+      }
+      ScalePosterior(opts.acoustic_weight * max_count_scale,
+                     &posterior);
+      // note: now, this_t == sum of posteriors.
+
+      int32 num_chunks = (mat.NumRows() - chunk_size + period) / period;
+      int32 final_chunk_size = mat.NumRows() - period * num_chunks;
+      bool imperfect_size = final_chunk_size != chunk_size - period;
+      Matrix<BaseFloat> ivectors(num_chunks + int32(imperfect_size), extractor.IvectorDim());
+      for (int32 i = 0; i < num_chunks; i++) {
+        Vector<BaseFloat> ivector(extractor.IvectorDim());
+        SubMatrix<BaseFloat> sub_mat(mat, i * period, chunk_size, 0, mat.NumCols());
+        IvectorExtract(extractor, utt, Matrix<BaseFloat>(sub_mat),
+		       std::vector<std::vector<std::pair<int32, BaseFloat> > >
+		       (&posterior[i * period], &posterior[i * period + chunk_size]),
+      		       &ivector, auxf_ptr);
+        ivectors.CopyRowFromVec(ivector, i);
+      }
+      // Handle the final chunk of irregular size
+      if (imperfect_size) {
+        Vector<BaseFloat> ivector(extractor.IvectorDim());
+        SubMatrix<BaseFloat> sub_mat(mat, num_chunks * period, final_chunk_size, 0, mat.NumCols());
+        IvectorExtract(extractor, utt, Matrix<BaseFloat>(sub_mat),
+         	       std::vector<std::vector<std::pair<int32, BaseFloat> > >
+         	       (&posterior[num_chunks * period], &posterior[num_chunks * period + final_chunk_size]),
+        	       &ivector, auxf_ptr);
+        ivectors.CopyRowFromVec(ivector, num_chunks);
+      }
```

Comment:
This is pretty good, but I think it will be easier to maintain if you get rid of the if (imperfect_size) { } block, and find a way to include that logic succinctly in the main loop. You can use the `std::min()` function to help you pick a `chunk_size` that fits within the feature matrix. For example, when you're taking the subset of the posteriors, you can just do something like the following on line 77:

`&posterior[i * period], &posterior[i * period + std::min(mat.NumRows(), i * period + chunk_size)]`

You can do something similar with to get the submatrix. 

Also, the variable `num_chunks`, is really `num_chunks - 1` in your code. This could confuse a reader of your code. It would be better to define it as `int32 num_chunks = (mat.NumRows() - chunk_size + period) / period + 1;` to avoid confusion.